media-placeholders/src/main/java/org/wordpress/aztec/placeholders/PlaceholderManager.kt

```diff
@@ -0,0 +1,338 @@
+            }
+        }
+        val targetPosition = aztecText.getElementPosition(predicate) ?: return
+
+        insertInPosition(aztecAttributes ?: return, targetPosition, currentPosition)
+    }
+
+    private fun insertInPosition(attrs: AztecAttributes, targetPosition: Int, currentPosition: Int? = null) {
+        validateAttributes(attrs)
+        val id = attrs.getValue(ID_ATTRIBUTE)
+        val uuid = attrs.getValue(UUID_ATTRIBUTE)
+        val type = attrs.getValue(TYPE_ATTRIBUTE)
+        val textViewLayout: Layout = aztecText.layout
+        val parentTextViewRect = Rect()
+        val targetLineOffset = getLineForOffset(targetPosition)
+        if (currentPosition != null) {
+            if (targetLineOffset != 0 && currentPosition == targetPosition) {
+                return
+            } else {
+                positionToId.removeAll {
+                    it.uuid == uuid
+                }
+            }
+        }
+        textViewLayout.getLineBounds(targetLineOffset, parentTextViewRect)
+
+        val parentTextViewLocation = intArrayOf(0, 0)
+        aztecText.getLocationOnScreen(parentTextViewLocation)
+        val parentTextViewTopAndBottomOffset = aztecText.scrollY + aztecText.compoundPaddingTop
+
+        parentTextViewRect.top += parentTextViewTopAndBottomOffset
+        parentTextViewRect.bottom += parentTextViewTopAndBottomOffset
+
+        var box = container.findViewWithTag<View>(uuid)
+        val exists = box != null
+        val drawer = drawers[type]!!
+        if (!exists) {
+            box = drawer.createView(container.context, id, attrs)
+        }
+        val params = FrameLayout.LayoutParams(
+                parentTextViewRect.right - parentTextViewRect.left - 20,
+                parentTextViewRect.bottom - parentTextViewRect.top - 20
+        )
+        val padding = 10
+        params.setMargins(parentTextViewRect.left + padding, parentTextViewRect.top + padding, parentTextViewRect.right - padding, parentTextViewRect.bottom - padding)
+        box.layoutParams = params
+        box.tag = uuid
+        box.setBackgroundColor(Color.TRANSPARENT)
+        positionToId.add(Placeholder(targetPosition, id, uuid))
+        if (!exists && box.parent == null) {
+            container.addView(box)
+            drawer.onViewCreated(box, id)
+        }
+    }
+
+    private fun getLineForOffset(offset: Int): Int {
+        var counter = 0
+        var index = 0
+        for (line in aztecText.text.split("\n")) {
+            counter += line.length + 1
+            if (counter > offset) {
+                break
+            }
+            index += 1
+        }
+        return index
+    }
+
+    private fun validateAttributes(attributes: AztecAttributes): Boolean {
+        if (!attributes.hasAttribute(UUID_ATTRIBUTE)) return false
+        if (!attributes.hasAttribute(TYPE_ATTRIBUTE)) return false
+        val type = attributes.getValue(TYPE_ATTRIBUTE)
+        return drawers[type] != null
```

Comment:
This could be simplified, but maybe it's more readable the way it is. I just wanted to play with it:

```suggestion
        return when {
            !attributes.hasAttribute(UUID_ATTRIBUTE) || !attributes.hasAttribute(TYPE_ATTRIBUTE) -> false
            else -> drawers[attributes.getValue(TYPE_ATTRIBUTE)] != null
        }
```

Reply:
Good point 👍, I've simplified it into one line